chore: contributing guide (#441)
algolia/algoliasearch-client-ruby: CONTRIBUTING.MD

<p align="center">
  <a href="https://www.algolia.com">
      <img alt="Algolia for Ruby" src="https://raw.githubusercontent.com/algolia/algoliasearch-client-common/master/banners/ruby.png" >
    </a>
</p>

Hello and welcome to the contributing guide for algolia gem. Thanks for considering participating in our project 🙇

If this guide does not contain what you are looking for and thus prevents you from contributing, don't hesitate to leave a message on the [community forum](https://discourse.algolia.com/) or to [open an issue](https://github.com/algolia/instantsearch.js/issues).

<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN TocDown TO UPDATE -->

> <!-- TocDown Begin -->

Author: it's not a gem unfortunately it's a script. I wasn't sure if its usage would fit anywhere in the project

  .rubocop.yml              << contains the rule used for the linter
  algolia.gemspec           << gemspec file
  CHANGELOG.md              << changelog file
  CONTRIBUTING.md           << this file
  Gemfile                   << dependencies needed to run the project locally
  Rakefile                  << defines the different tasks to lint/test the project
  README.md                 << the introduction of the project
```

### The lib/algolia folder

```
▸ lib/algolia/
    ▸ config/       << the configurations associated with each clients 
    ▸ enums/        << enumerables used accross the lib
    ▸ http/         << the http layer logic
    ▸ iterators/    << the iterators used for the browsing methods
    ▸ responses/    << the waitable responses
    ▸ transport/    << the transport layer logic
```

## Tests

Our unit and integration tests are written with [Minitest](https://github.com/seattlerb/minitest), in the `test` syntax.

To run all the tests once:

```sh
bundle exec rake test:all
```

To run only the unit tests

```sh
bundle exec rake test:unit
```

To run only the integration tests

```sh
bundle exec rake test:integration
```

> We ask that for each fix or feature submitted, you add at least once test demonstrating its behaviour. As you will need to use your own Algolia

Reviewer: ```suggestion
We ask that for each fix or feature submitted, you add at least one test demonstrating its behaviour. As you will need to use your own Algolia
```

<p align="center">
  <a href="https://www.algolia.com">
      <img alt="Algolia for Ruby" src="https://raw.githubusercontent.com/algolia/algoliasearch-client-common/master/banners/ruby.png" >
    </a>
</p>

Hello and welcome to the contributing guide for algolia gem. Thanks for considering participating in our project 🙇

> If this guide does not contain what you are looking for and thus prevents you from contributing, don't hesitate to leave a message on the [community forum](https://discourse.algolia.com/) or to [open an issue](https://github.com/algolia/instantsearch.js/issues).

Reviewer 2: ```suggestion
If this guide does not contain what you are looking for and thus prevents you from contributing, don't hesitate to leave a message on the [community forum](https://discourse.algolia.com/) or to [open an issue](https://github.com/algolia/algoliasearch-client-ruby/issues).
```

<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN TocDown TO UPDATE -->
<!-- TocDown Begin -->
* [Reporting an issue](#reporting-an-issue)
* [The code contribution process](#the-code-contribution-process)
* [Commit conventions](#commit-conventions)
* [Branch organization](#branch-organization)
* [Requirements](#requirements)
* [Code structure](#code-structure)
  * [The source algolia folder](#the-source-algolia-folder)
* [Tests](#tests)
* [Linting](#linting)
<!-- TocDown End -->

## Reporting an issue

> Opening an issue is very effective way to contribute because many users might also be impacted. We'll make sure to fix it quickly if it's technically feasible and doesn't have important side effects for other users.

Reviewer 2: ```suggestion
Opening an issue is very effective way to contribute because other users might also be impacted. We'll make sure to fix it quickly if it's technically feasible and doesn't have important side effects for other users.
```

Before reporting an issue, first check that there is not an already open issue for the same topic using the [issues page](https://github.com/algolia/algoliasearch-client-ruby/issues). Don't hesitate to thumb up an issue that corresponds to the problem you have.

> Another element that will help us go faster at solving the issue is to provide a reproducible test case.

Reviewer 2: ```suggestion
It would be very helpful if you're able to add a test case that reproduces the issue. This could help us solve the issue faster.
```

## The code contribution process

The algolia gem is developed in Ruby ≥ 2.2.

For any code contribution, you need to:

- Fork and clone the project
- Create a new branch for what you want to solve (fix/_issue-number_, feat/_name-of-the-feature_)
- Make your changes
- Open a pull request

Depending on what you're working on, you might consider different [base branches](#branch-organization).

Then:

- Peer review of the pull request (by at least one of the core contributors)
- Automatic checks ([tests](#tests), [commits](#commit-conventions), [linters](#linting))

> - When everything is green, your contribution is merged 🚀

Reviewer 2: The way we phrase it now makes it seem like it's an automatic process I think
```suggestion
- When everything is green, one of the core contributors will merge your contribution 🚀
```

## Code structure

Here are the main files and folders of the project

```
▸ lib/algolia/              << standalone clients and helpers classes
▸ lib/test/                 << gathers the unit and integration tests
  .rubocop.yml              << contains the rule used for the linter
  algolia.gemspec           << gemspec file
  CHANGELOG.md              << changelog file
  CONTRIBUTING.md           << this file
  Gemfile                   << dependencies needed to run the project locally
  Rakefile                  << defines the different tasks to lint/test the project
  README.md                 << the introduction of the project
```

### The lib/algolia folder

```
▸ lib/algolia/
    ▸ config/       << the configurations associated with each clients 
    ▸ enums/        << enumerables used accross the lib
    ▸ http/         << the http layer logic
    ▸ iterators/    << the iterators used for the browsing methods
    ▸ responses/    << the waitable responses
    ▸ transport/    << the transport layer logic
```

## Tests

Our unit and integration tests are written with [Minitest](https://github.com/seattlerb/minitest), in the `test` syntax.

> To run all the tests once:

Reviewer 2: ```suggestion
To run all the tests at once:
```

```sh
bundle exec rake test:all
```

> To run only the unit tests

Reviewer 2: ```suggestion
To run only the unit tests:
```

```sh
bundle exec rake test:unit
```

> To run only the integration tests

Reviewer 2: ```suggestion
To run only the integration tests:
```

```sh
bundle exec rake test:integration
```

We ask that for each fix or feature submitted, you add at least one test demonstrating its behaviour. As you will need to use your own Algolia
credentials to run them, we advise you to keep them short and use a small dataset, as well as using a mock requester anytime it's possible. If you 

> encounter huge data overload because of testing, please reach out to [Support](support@algolia.com).

Reviewer 2: ```suggestion
encounter huge data overload because of testing, please reach out to [our support team](support@algolia.com).
```

> We ask that for each fix or feature submitted, you add at least one test demonstrating its behaviour. As you will need to use your own Algolia
> credentials to run them, we advise you to keep them short and use a small dataset, as well as using a mock requester anytime it's possible. If you 
> encounter huge data overload because of testing, please reach out to [our support team](support@algolia.com).

Reviewer 2: Can you please add a link on the `using a mock requester` to link to the docs on how to do that?
```suggestion
We ask that for each fix or feature submitted, you add at least one test demonstrating its behaviour. As you will need to use your own Algolia credentials to run them, we advise you to keep them short and use a small dataset, as well as using a mock requester anytime it's possible. If you encounter huge data overload because of testing, please reach out to [our support team](support@algolia.com).
```

## Linting

> Linters are static checkers for code. They help us maintain a consistent code base.

Reviewer 2: ```suggestion
Linters help us maintain a consistent code base.
```
